File: 0022-generate-parentheses/0022-generate-parentheses.py
class Solution:
    def generateParenthesis(self, n: int) -> list[str]:
        result = []
        
        def backtrack(open_count: int, close_count: int, current_string: str):
            # Base Case: Valid combination reached
            if len(current_string) == 2 * n:
                result.append(current_string)
                return
            
            # Rule 1: Add open parenthesis if available
            if open_count < n:
                backtrack(open_count + 1, close_count, current_string + "(")
                
            # Rule 2: Add close parenthesis if it matches an open one
            if close_count < open_count:
                backtrack(open_count, close_count + 1, current_string + ")")
                
        backtrack(0, 0, "")
        return result

File: 0022-generate-parentheses/0022-generate-parentheses.java
class Solution {
    public List<String> generateParenthesis(int n) {
        List<String> result = new ArrayList<>();

        generate(0, 0, new StringBuilder(), n, result);

        return result;
    }

    public void generate(int countOpen, int countClose, StringBuilder sb, int n, List<String> result) {
        if(sb.length() == 2*n) {
            result.add(sb.toString());
            return;
        }

        if(countOpen < n) {
            generate(countOpen + 1, countClose, sb.append("("), n, result);
            sb.deleteCharAt(sb.length()-1);
        }
        
        if(countClose < countOpen) {
            generate(countOpen, countClose + 1, sb.append(")"), n, result);
            sb.deleteCharAt(sb.length()-1);
        }
    }
}